src/ppm.rs

```diff
@@ -0,0 +1,473 @@
+            hasher.update(&r.into());
+        }
+
+        helper[i][JOINT_RAND_HINT].copy_from_slice(hasher.finalize().as_bytes());
+        for j in 0..SEED_LENGTH {
+            joint_rand_seed[j] ^= helper[i][JOINT_RAND_HINT][j];
+        }
+    }
+
+    let mut leader_blind = vec![0; SEED_LENGTH];
+    getrandom(&mut leader_blind)?;
+
+    let mut hasher = blake3::Hasher::new();
+    hasher.update(&[0]); // aggregator_id
+    hasher.update(&leader_blind);
+    for j in 0..input_len {
+        hasher.update(&leader_input_share[j].into());
+    }
+
+    let mut leader_joint_rand_seed_hint = vec![0; SEED_LENGTH];
+    leader_joint_rand_seed_hint.copy_from_slice(hasher.finalize().as_bytes());
+    for j in 0..SEED_LENGTH {
+        joint_rand_seed[j] ^= leader_joint_rand_seed_hint[j];
+    }
+
+    // Run the proof-generation algorithm.
+    let mut prng = Prng::from_seed(&joint_rand_seed)?;
+    let joint_rand: Vec<F> = (0..input.joint_rand_len())
+        .map(|_| prng.next().unwrap())
+        .collect();
+    let prove_rand: Vec<F> = rand(input.prove_rand_len())?;
+    let proof = prove(input, &prove_rand, &joint_rand)?;
+
+    // Generate the proof shares and finalize the joint randomness seed hints.
+    let proof_len = proof.as_slice().len();
+    let mut leader_proof_share = proof.data;
+    for i in 0..num_shares - 1 {
+        getrandom(&mut helper[i][PROOF_SHARE])?;
+
+        let mut prng = Prng::from_seed(&helper[i][PROOF_SHARE])?;
+        for j in 0..proof_len {
+            let r = prng.next().unwrap();
+            leader_proof_share[j] -= r;
+        }
+
+        for j in 0..SEED_LENGTH {
+            helper[i][JOINT_RAND_HINT][j] ^= joint_rand_seed[j];
+        }
+    }
+
+    for j in 0..SEED_LENGTH {
+        leader_joint_rand_seed_hint[j] ^= joint_rand_seed[j];
+    }
+
+    // Prepare the output messages.
+    let mut aggregator_id = 0;
+    let mut msgs = Vec::with_capacity(num_shares);
+    msgs.push(PrioUpload {
+        aggregator_id,
+        input_share: PrioShare::Leader(leader_input_share),
+        proof_share: PrioShare::Leader(leader_proof_share),
+        joint_rand_seed_hint: leader_joint_rand_seed_hint,
+        blind: leader_blind,
+    });
+
+    for row in helper.into_iter() {
+        aggregator_id += 1;
+        let mut col = row.into_iter();
+        msgs.push(PrioUpload {
+            aggregator_id,
+            input_share: PrioShare::Helper(col.next().unwrap(), input_len),
+            proof_share: PrioShare::Helper(col.next().unwrap(), proof_len),
+            joint_rand_seed_hint: col.next().unwrap(),
+            blind: col.next().unwrap(),
+        });
+    }
+
+    Ok(msgs)
+}
+
+/// The message sent by an aggregator to every other aggregator. This is the final message of the
+/// input-validation protocol.
+#[derive(Clone)]
+pub struct PrioVerifier<F: FieldElement> {
+    /// The aggregator's share of the verifier message.
+    pub verifier_share: Vec<F>,
```

Comment:
```suggestion
    pub verifier_share: Verifier<F>,
```
`Verifier<F>` is basically just `Vec<F>` anyway, but it's more clear -- so many different things are represented as `Vec<F>` or `&[F]` in this library that it's easy to get confused or pass the wrong variable somewhere.

Reply:
I agree in the case of `Share` because it's useful for it to be flexible, but here we can gain some clarity without loss of usefulness of this struct. That said we then end up with a collision between `ppm::Verifier` and `pcp::Verifier`. Maybe this should be `ppm::VerifierMessage`?